### PR TITLE
Lighting optimizations

### LINKED ISSUE #25629: Lighting objects needlessly uses expensive color matrices for fully lit/fully dark tiles.
For whatever reason Byond's implementation of color matrices is very inefficient on older computers (that or it's inefficient everywhere but just harder to ignore on older computers). Costly client rendering seemingly occurs every time a matrix colored object enters the screen. Considering the fact that there's up to 215 lighting objects on the screen at once this cost can really add up.

It is impossible not to use color matrices for colored/gradient light objects. However for uncolored fully lit (read: transparent) and fully dark (read: black) lighting objects this expensive matrix operation could be replaced with a much simpler flat color/alpha that would save a lot of client cost without changing what the lighting actually looks like at all.

I've been chipping away at why the new lighting runs so terribly on my local machine and this bug represents my findings. I am currently attempting to figure out a nice and cheap way to solve this but if someone else with a better understanding of the system beats me to the punch I wouldn't mind it at all.

### PR DESCRIPTION
**Pull request by Cyberboss**
Fixes #25629

@PJB3005 for review. Make sure my understanding of the color matrix is right. Mainly this line https://github.com/tgstation/tgstation/pull/25654/files#diff-7528eb00f20f9d0ea0e4bb77631220f1R104

### CODE REVIEW

**Comment by Incoming5643** [comment]
The problem I was implying was that any client side optimization here was by design a deoptimization server side due to added checks.

I'm hopeful to find a way to do this stuff without giant 20 var `&&` chains

File: code/modules/lighting/lighting_object.dm [hunk 1]
```diff
@@ -75,21 +75,47 @@
...
 	#else
 	// Because of floating points™️, it won't even be a flat 0.
 	// This number is mostly arbitrary.
-	luminosity = max > 1e-6
+	var/set_luminosity = max > 1e-6
 	#endif
 
+	if(set_luminosity && 
+		//check if the tile is fully lit
+		!(rr == 255 && rg == 255 && rb == 255 && gr == 255 && gg == 255 && gb == 255 && br == 255 && bg == 255 && bb == 255 && ar == 255 && ag == 255 && ab == 255)
```
Comment (PJB3005) [inline code/modules/lighting/lighting_object.dm]: This indeed does not work.

The values are from 0-1.

**Comment by Cyberboss** [comment]
@Incoming5643 The if will have near zero performance impact on the server. Checking proc vars against constants is stupid fast and, given the structure of the check, if it's gonna fail, it'll fail fast.

**Comment by Incoming5643** [comment]
To be clear this matrix represents a full bright lighting object:
`1 1 1 0`
`1 1 1 0`
`1 1 1 0`
`1 1 1 0`
`0 0 0 1`

and THIS matrix represents a full dark lighting object:

A = 1.78814e-007
B = -5.96046e-008

`A A A 0` 
`A A A 0`
`B B B 0`
`B B B 0`
`0 0 0 1`

If full dark tiles weren't handed with matrix coloring the infinitesimal adds might not be needed anymore, they were only added to get around a byond bug.

**Comment by Cyberboss** [comment]
The darkness is determined separately. It just checks if a turf is luminous or not

**Comment by YotaXP** [comment]
Would `rr + rg + rb + gr + ... + ag + ab == 12` be faster?  Or maybe `> 11.99` because floating point.

**Comment by Cyberboss** [comment]
Wow you are so right
```
/world
	loop_checks = 0

/world/New()
	var/test1 = 1
	var/test2 = 1
	var/test3 = 1
	var/test4 = 1
	var/test5 = 1
	var/test6 = 1
	var/test7 = 1
	var/test8 = 1
	var/test9 = 1
	var/test10 = 1
	var/test11 = 1
	var/test12 = 1 

	var/start_time
	var/end_time
	start_time = timeofday

	for(var/I in 1 to 10000000)
		if(test1 + test2 + test3 + test4 + test5 + test6 + test7 + test8 + test9 + test10 + test11 + test12 > 11.99)
			Whatever()

	end_time = timeofday

	log << "addition took [(end_time-start_time)/10]s"

	start_time = timeofday

	for(var/I in 1 to 10000000)
		if(test1 == 1 && test2 == 1 && test3 == 1 && test4 == 1 && test5 == 1 && test6 == 1 && test7 == 1 && test8 == 1 && test9 == 1 && test10 == 1 && test11 == 1 && test12 == 1)
			Whatever()

	end_time = timeofday

	log << "comparison took [(end_time-start_time)/10]s"

	del(src)

/proc/Whatever()
```
Result:
```
addition took 7s
comparison took 11.3s
```

**Comment by Cyberboss** [comment]
GRANTED:
There is the (frequent) case of failing fast

I set `test2 = 0.5` for this test and got:
```
addition took 5s
comparison took 1.9s
```

So I'm going to ask a maintainer for their opinion here

**Comment by Incoming5643** [comment]
Well to be fair that's the worst case situation for the comparison, as you said most of the time that `&&` chain will short circuit somewhere before the entire chain resolves. Addition will always take the same amount of time.

Edit: dang it you beat me to it

**Comment by Incoming5643** [comment]
Personal testing shows me that comparison is still faster so long as the short circuiting happens at or before the sixth or so entry. In the current check that gets you through looking through two out of four corners.

I suggest formatting it instead like this:
`!(rr == 1 && gr == 1 && br == 1 && ar == 1 && rb == 1 && gb == 1 ...`

Because that allows you to touch all corners quicker in a way that usually will be enough to properly show that tile isn't fullbright.

---

Also I don't think the dark tile check works because `set_luminosity = max > 1e-6` intends to trick the system into not letting luminosity ever be exactly 0, which it would need to be for `if(set_luminosity)` to fail. It'd have to be something like `if(set_luminosity > 0.0001)` or something unfortunately

**Comment by Cyberboss** [comment]
`set_luminosity` will always be 1 or 0. `max > 1e-6` is a boolean statement

**Comment by RemieRichards** [comment]
@Incoming5643 set luminosity is a bool you twit.

**Comment by Incoming5643** [comment]
🤡

File: code/modules/lighting/lighting_object.dm [hunk 2]
```diff
@@ -75,21 +75,47 @@
...
 	#else
 	// Because of floating points™️, it won't even be a flat 0.
 	// This number is mostly arbitrary.
-	luminosity = max > 1e-6
+	var/set_luminosity = max > 1e-6
 	#endif
 
+	if(set_luminosity && \
+		//check if the tile is fully lit
+		!(rr == 1 && gr == 1 && br == 1 && ar == 1 && rg == 1 && gg == 1 && bg == 1 && ag == 1 && rb == 1 && gb == 1 && bb == 1 && ab == 1)\
```
Comment (optimumtact) [inline code/modules/lighting/lighting_object.dm]: You might eke a bit more speed if you & these together

quick tests in irc show taht  !dm var/x = 1&0.9&1&1&0.5; world.log << x;; returns 0 and all 1's return 1

Comment (optimumtact) [inline code/modules/lighting/lighting_object.dm]: then honestly you can probably & with set_luminosty as well

**Comment by Incoming5643** [comment]
ran the most recent version of this code locally to see how it performed.

The good news: It ran so much smoother for me!

The bad news: It looked like this
scrnshot1

**Comment by optimumtact** [comment]
I dont' see any issue :grin:

**Pull request closed without merging**